Track active hosts count and enforce limit

mna opened the pull request:
#5520 

# Checklist for submitter

If some of the following don't apply, delete the relevant line.

- [x] Changes file added for user-visible changes (in `changes/` and/or `orbit/changes/`).
- [x] Added/updated tests

mna commented on cmd/fleet/serve.go:
```diff
@@ -291,6 +292,13 @@ the way that the Fleet server works.
 
 			ds = cached_mysql.New(ds)
+			var dsOpts []mysqlredis.Option
+			// TODO(mna): clarify condition to add the host count check for hard enrollment limit.
```
@chiiph Unclear to me how to enable the enforcement of the hosts limit, how do I detect that this is running for Fleet Demo?

chiiph replied:
We can have a new option/env var `enforce_license_limits` the defaults to false.

chiiph replied:
Not without some help from a variable/param being set.

mna commented on server/datastore/mysql/schema.sql:
The auto-generated SQL schema had changes for me, I _think_ this is something @michalnicp added recently, does that look correct to you?

michalnicp replied:
This looks correct according to the migration in `server/datastore/mysql/migrations/tables/20220526123327_RenameCVEScoresToCVEMeta.go`. I may have forgotten to commit the changes to `schema.sql` on one of my recent prs.

mna commented on server/datastore/mysqlredis/hosts.go:
```diff
@@ -0,0 +1,140 @@
+// For additional safety, we could check the DB if we're about to reject an
+// enrollment: if so, check the COUNT(*) in the DB, and if it matches or is
+// above, reject, if it doesn't match load the IDs and replace the redis SET
+// with those.
```
The implementation is quite straightforward for now - just track the host IDs in a Redis set whenever one is enrolled/created, and remove it when deleted (explicitly or via cleanup tasks).

What I'd suggest is to also implement the cron job to re-sync in case the redis set gets out of sync (or redis gets restarted for some reason, etc.). I would not add the "additional safety" thing I mention above, unless you feel it's crucial to be absolutely right. / @chiiph

michalnicp replied:
I would recommend implementing this now. Redis restarting should be an expected event. If it restarts at any point after any hosts get added, then we would never be able to recover ie we would only count hosts after the **last restart** of redis towards the enroll limit.

mna replied:
Yes, that's what I recommend too.

mna commented on server/datastore/mysqlredis/hosts_test.go:
```diff
@@ -0,0 +1,136 @@
+		require.Contains(t, err.Error(), "maximum number of hosts")
+		require.False(t, ds.EnrollHostFuncInvoked)
+
+		_, _, _ = h6, h7, h8 // unused but makes test cases clearer
```
I can remove this and the corresponding vars if y'all hate it, but I feel it helps keep track of the hosts in the test.

mna commented on server/live_query/redis_live_query.go:
```diff
@@ -333,17 +333,27 @@ func mapBitfield(hostIDs []uint) []byte {
+	// usage of memory would be true even if there was only one host selected in
+	// the query, should that host be one of the high IDs.  Something to keep in
+	// mind if at some point we have reports of unexpectedly large redis memory
+	// usage, as that storage is repeated for each live query.
```
Unrelated to the change per se, but I took a closer look at the set size for this PR and wanted to update this so that the comment reflects the actual behaviour better.

mna commented on server/datastore/mysqlredis/hosts.go:
```diff
@@ -0,0 +1,140 @@
+	return h, err
+}
+
+func (d *datastore) EnrollHost(ctx context.Context, osqueryHostID, nodeKey string, teamID *uint, cooldown time.Duration) (*fleet.Host, error) {
```
I'll see how that would look. I agree about the principle but I'm afraid this might end up being significantly more complex, since the service needs to be aware of whether the enforcement is enabled or not (instead of being abstracted away in the Datastore from the moment `serve` is called), and it needs a Datastore method to call to get the count - which (awkwardly) would be implemented only by the redis wrapper datastore, being a no-op elsewhere - just so it can test the condition and fail if it's over the limit.

(arguably the non-redis-wrapper datastore could return `0, nil` so that the service does not need to know if enforcement is enabled, but then the datastore method is plain wrong if enforcement is not enabled, e.g. calling `Datastore.GetEnrolledHostsCount()` returns a number that may or may not be right depending on whether the check is enabled)

michalnicp replied:
> the service needs to be aware of whether the enforcement is enabled or not (instead of being abstracted away in the Datastore from the moment fleet serve is called)

The `Service` already has access to `FleetConfig`, so this would be an if check. We would need to pass the `deviceCount` though.

> needs a Datastore method to call to get the count - which (awkwardly) would be implemented only by the redis wrapper datastore, being a no-op elsewhere

We don't have to use the wrapping method here. We could just pass a redis connection to the datastore in `mysql.New(...)`. `GetEnrolledHostsCount` would only be called by the service if it needs to. Same goes for incrementing the enrolled hosts count. The service would be responsible for calling these datastore methods appropriately.

mna replied:
> The Service already has access to FleetConfig, so this would be an if check. We would need to pass the deviceCount though.

True, it also has the license info so it does have the `deviceCount`.

> We could just pass a redis connection to the datastore in mysql.New(...).

This I'm not convinced though. I'd still use the wrapping datastore so the redis access is only done in the redis wrapper.

> Same goes for incrementing the enrolled hosts count. The service would be responsible for calling these datastore methods appropriately.

I disagree that this is cleaner than the wrapped Datastore, I think it's much nicer to have the wrapper take care transparently of keeping track of the counts rather than add this in all related Service methods (plus the cleanup calls are not even triggered by a service method, it's in a cron job). It's not only nicer, it's safer - if the datastore methods end up being called elsewhere, they will still maintain proper counts. I agree that the actual count validation check makes sense in the `Service`, but not maintaining the redis set IMHO.

mna commented on server/datastore/mysql/hosts.go:
```diff
@@ -546,17 +546,31 @@ func (ds *Datastore) CountHosts(ctx context.Context, filter fleet.TeamFilter, op
+		return nil, ctxerr.Wrap(ctx, err, "load incoming hosts to cleanup")
+	}
+
+	cleanupHosts := `
```
This would potentially send a lot more data though (and I believe there is some limit to the number of values one can pass in a `IN (?)` condition). The statements are guaranteed to affect the same IDs, given that the time reference is static and the condition is based on the `created_at` timestamp (unless we'd start creating hosts in the past).

roperzh replied:
gotcha, thanks for clarifying!

juan-fdz-hawa commented on server/datastore/mysql/hosts.go:
```diff
@@ -1077,39 +1091,32 @@ func (ds *Datastore) CleanupExpiredHosts(ctx context.Context) error {
 
-	rows, err := ds.writer.QueryContext(
+	var ids []uint
+	err = ds.writer.SelectContext(
```
nit: `ds.writer` -> `ds.reader`

juan-fdz-hawa replied:
nit: `ds.writer` -> `ds.reader`

juan-fdz-hawa commented on server/datastore/mysql/hosts.go:
```diff
@@ -546,17 +546,31 @@ func (ds *Datastore) CountHosts(ctx context.Context, filter fleet.TeamFilter, op
+    WHERE
+      hostname = '' AND
+      osquery_version = '' AND
+      created_at < (? - INTERVAL 5 MINUTE)`
```
I guess this is a little bit paranoid but, the `created_at < (....)` here and the `created_at < (...)` used in the `DELETE` below can potentially return different results.

juan-fdz-hawa commented on server/datastore/mysql/hosts.go:
```diff
@@ -546,17 +546,31 @@ func (ds *Datastore) CountHosts(ctx context.Context, filter fleet.TeamFilter, op
+    WHERE
+      hostname = '' AND
+      osquery_version = '' AND
+      created_at < (? - INTERVAL 5 MINUTE)`
```
I guess this is a little bit paranoid but, the `created_at < (....)` here and the `created_at < (...)` used in the `DELETE` below can potentially return different results.

mna replied:
In practice it can't - it uses the same WHERE clause with a fixed timestamp using the `created_at` column as condition, so it cannot create new rows 5+ minutes in the past.

juan-fdz-hawa replied:
Thanks @mna, you are right (seems like I need a little bit more coffee)

mna commented:
The PR has been updated and comments addressed. In particular, it addresses those comments:

* https://github.com/fleetdm/fleet/pull/6099#discussion_r890287228 (adding a cron job to ensure at regular intervals that the redis set is in sync with the mysql source of truth)
* https://github.com/fleetdm/fleet/pull/6099#discussion_r890322729 (move the check to enroll a new host in the service layer, so that the validation is more explicit)

Note that the `Service` struct has a lot of redis-related fields that implement some methods based on the redis db (e.g. QueryResultStore, LiveQueryStore, SessionStore, FailingPolicySet and now EnrollHostLimiter). Those are different fields, but essentially all the same core value stored multiple times (a `fleet.RedisPool` wrapped in a struct just to implement specific methods using that pool). What I'd suggest is to do something similar to what we did in the past with the mysql Datastore (which was multiple different interfaces all fulfilled by the same struct) and to have a single value satisfy all the various redis-backed interfaces so that we can stop the constant addition of fields to the service for such common cases. If y'all are ok with that approach (no need to go in all the details here), I'll create a ticket so it can be addressed later on (not in this PR).

Desmi-Dizney commented:
Editor pass completed on: 
- https://github.com/fleetdm/fleet/pull/6224
